File: traditional_comm_py/traditional_comm/transport.py
from __future__ import annotations

from abc import ABC, abstractmethod
import time


class TransportAdapter(ABC):
    def health_check(self) -> dict:
        return {"online": True, "transport": type(self).__name__}

    @abstractmethod
    def send_payload(self, payload: bytes, metadata: dict | None = None) -> tuple[bytes, dict]:
        raise NotImplementedError


class LoopbackTransport(TransportAdapter):
    def __init__(self, chunk_size: int = 16 * 1024, delay_ms: float = 0, loss_rate: float = 0):
        if chunk_size <= 0 or int(chunk_size) != chunk_size:
            raise ValueError("chunk_size must be a positive integer")
        if delay_ms < 0 or not 0 <= loss_rate < 1:
            raise ValueError("delay_ms must be >= 0 and loss_rate must be in [0, 1)")
        self.chunk_size = int(chunk_size)
        self.delay_ms = float(delay_ms)
        self.loss_rate = float(loss_rate)

    def send_payload(self, payload: bytes, metadata: dict | None = None) -> tuple[bytes, dict]:
        metadata = metadata or {}
        start = time.perf_counter()
        chunks: list[bytes] = []
        lost_chunks = 0
        chunk_count = (len(payload) + self.chunk_size - 1) // self.chunk_size
        for offset in range(0, len(payload), self.chunk_size):
            chunk = payload[offset : offset + self.chunk_size]
            if self.delay_ms:
                time.sleep(self.delay_ms / 1000)
            # A zero loss rate is deterministic; nonzero loss is intentionally a smoke-test simulation.
            import random
            if self.loss_rate and random.random() < self.loss_rate:
                lost_chunks += 1
                continue
            chunks.append(bytes(chunk))
        received = b"".join(chunks)
        elapsed_ms = (time.perf_counter() - start) * 1000
        seconds = max(elapsed_ms / 1000, 1e-9)
        average_mbps = len(received) * 8 / seconds / 1_000_000
        stats = {
            "transport": "loopback",
            "run_id": metadata.get("run_id"),
            "sent_bytes": len(payload),
            "received_bytes": len(received),
            "chunk_size": self.chunk_size,
            "chunk_count": chunk_count,
            "lost_chunks": lost_chunks,
            "loss_rate": self.loss_rate,
            "delay_ms_per_chunk": self.delay_ms,
            "elapsed_ms": round(elapsed_ms, 3),
            "average_throughput_mbps": round(average_mbps, 3),
            "peak_throughput_mbps": round(average_mbps, 3),
        }
        return received, stats


class Task3Transport(TransportAdapter):
    """Adapter seam for the future task-3 communication link."""

    def __init__(self, implementation):
        if implementation is None or not hasattr(implementation, "send_payload"):
            raise ValueError("Task3Transport requires an implementation with send_payload()")
        self.implementation = implementation

    def health_check(self) -> dict:
        if hasattr(self.implementation, "health_check"):
            return self.implementation.health_check()
        return {"online": True, "transport": "task3-adapter"}

    def send_payload(self, payload: bytes, metadata: dict | None = None) -> tuple[bytes, dict]:
        return self.implementation.send_payload(payload, metadata or {})

File: traditional_comm/src/transport.js
const { setTimeout: delay } = require('node:timers/promises');

class TransportAdapter {
  async healthCheck() {
    return { online: true, transport: this.constructor.name };
  }

  async sendPayload() {
    throw new Error('TransportAdapter.sendPayload() must be implemented');
  }
}

class LoopbackTransport extends TransportAdapter {
  constructor({ chunkSize = 16 * 1024, delayMs = 0, lossRate = 0 } = {}) {
    super();
    if (!Number.isInteger(chunkSize) || chunkSize <= 0) {
      throw new Error('chunkSize must be a positive integer');
    }
    if (delayMs < 0 || lossRate < 0 || lossRate >= 1) {
      throw new Error('delayMs must be >= 0 and lossRate must be in [0, 1)');
    }
    this.chunkSize = chunkSize;
    this.delayMs = delayMs;
    this.lossRate = lossRate;
  }

  async sendPayload(payload, metadata = {}) {
    if (!Buffer.isBuffer(payload)) {
      throw new TypeError('LoopbackTransport payload must be a Buffer');
    }

    const start = process.hrtime.bigint();
    const chunks = [];
    let lostChunks = 0;

    for (let offset = 0; offset < payload.length; offset += this.chunkSize) {
      const chunk = payload.subarray(offset, Math.min(offset + this.chunkSize, payload.length));
      if (this.delayMs > 0) await delay(this.delayMs);
      if (this.lossRate > 0 && Math.random() < this.lossRate) {
        lostChunks += 1;
        continue;
      }
      chunks.push(Buffer.from(chunk));
    }

    const received = Buffer.concat(chunks);
    const elapsedMs = Number(process.hrtime.bigint() - start) / 1e6;
    const seconds = Math.max(elapsedMs / 1000, 1e-9);
    const averageMbps = received.length * 8 / seconds / 1e6;
    const stats = {
      transport: 'loopback',
      run_id: metadata.run_id || null,
      sent_bytes: payload.length,
      received_bytes: received.length,
      chunk_size: this.chunkSize,
      chunk_count: Math.ceil(payload.length / this.chunkSize),
      lost_chunks: lostChunks,
      loss_rate: this.lossRate,
      delay_ms_per_chunk: this.delayMs,
      elapsed_ms: Number(elapsedMs.toFixed(3)),
      average_throughput_mbps: Number(averageMbps.toFixed(3)),
      peak_throughput_mbps: Number(averageMbps.toFixed(3)),
    };
    return { payload: received, stats };
  }
}

class Task3Transport extends TransportAdapter {
  constructor(implementation) {
    super();
    if (!implementation || typeof implementation.sendPayload !== 'function') {
      throw new Error('Task3Transport requires an implementation with sendPayload()');
    }
    this.implementation = implementation;
  }

  async healthCheck() {
    if (typeof this.implementation.healthCheck === 'function') {
      return this.implementation.healthCheck();
    }
    return { online: true, transport: 'task3-adapter' };
  }

  async sendPayload(payload, metadata = {}) {
    return this.implementation.sendPayload(payload, metadata);
  }
}

module.exports = {
  TransportAdapter,
  LoopbackTransport,
  Task3Transport,
};
